concepts/eventing-code-guidelines.md

```diff
@@ -0,0 +1,1155 @@
+
+This guide shall be publicly available so that it can be referenced/used when reviewing pull requests. The agreements of this guide shall not affect `outside collaborators`. If a PR does not align with our guidelines, we should accept the PR as it is and apply the guidelines ourselves in a follow-up PR in order to not block external contributors. Of course, also external PRs shall meet our quality standards but we should not perform too much nit picking.
+
+For now the guidelines shall be applied to new code (PRs). There is no need to rewrite our entire code base now. This will gradually happen over time. As soon as the code reflects our guidelines, the need to look into this guide will decrease more and more.
+
+## Table of contents
+
+<!-- generate me with markdown-toc 
+```bash
+// source: https://github.com/jonschlinkert/markdown-toc
+markdown-toc -i --maxdepth 2 eventing-code-guidelines.md
+Do NOT TOUCH anything between the toc comments because this is used as a `marker` where to place the toc for markdown-toc.
+```
+-->
+
+<!-- toc -->
+
+- [Recommended Libraries](#recommended-libraries)
+- [Documentation Guidelines](#documentation-guidelines)
+- [Dockerfile Guidelines](#dockerfile-guidelines)
+- [Coding Guidelines](#coding-guidelines)
+- [Testing Pyramid in Kyma](#testing-pyramid-in-kyma)
+- [Testing Guidelines](#testing-guidelines)
+
+<!-- tocstop -->
+
+## Recommended Libraries
+
+The goal of this section is to describe the `desired choice` of `libraries` for testing, logging, etc.
+
+### Testing
+
+**Testing frameworks**:
+
+Use [`t.Testing`](https://pkg.go.dev/testing) for `unit` tests and for controller integration tests.
+
+<!-- voting options for testing framework in controller integration tests:
+1. testing.T // proposed by guidelines
+1. [ginkgo](https://github.com/onsi/ginkgo)
+-->
+
+**Assertion librariers**:
+
+<!-- voting options for assertion libraries in unit tests:
+Right now we use both: testing.T and gomega. We should reach a consens here to have consistency.
+
+1. testing.T
+2. [stretchr/testify](https://github.com/stretchr/testify)
+3. [onsi/gomega](https://github.com/onsi/gomega)
+-->
+
+Use [`gomega`](https://github.com/onsi/gomega) for writing test `assertions` in controller integration tests. Do not use [stretchr/testify](https://github.com/stretchr/testify) unless you have a reason for it.
+
+**Mocking Libraries**:
+
+Use [stretchr/testify/mock](https://github.com/stretchr/testify#mock-package) for generating mocks or create your own mock by implementing the corresponding interface.
+
+<!-- voting options:
+1. [stretchr/testify](https://github.com/stretchr/testify#mock-package) // proposed by guidelines and already used by eventing-controller
+2. gomock aka [golang/mock](https://github.com/golang/mock)
+3. [vektra/mockery](https://github.com/vektra/mockery)
```

Comment:
This mockery tool generates mocks using the `stretchr/testify` mock package.

Reply:
I used the mockery in my unit test [PR](https://github.com/kyma-project/kyma/pull/13315/files) and it was very easy and served completely my purposes

Reply:
@raypinto I see. Thanks for the explanation 😃 I will combine these two entries then.